RecoLocalCalo/HcalRecAlgos/plugins/HcalChannelPropertiesEP.cc

```diff
@@ -0,0 +1,117 @@
+#include "FWCore/Framework/interface/ESProducer.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "CalibFormats/HcalObjects/interface/HcalDbService.h"
+#include "CalibFormats/HcalObjects/interface/HcalDbRecord.h"
+
+#include "DataFormats/HcalDetId/interface/HcalGenericDetId.h"
+#include "DataFormats/HcalDetId/interface/HcalSubdetector.h"
+
+#include "CondFormats/HcalObjects/interface/HcalRecoParams.h"
+#include "CondFormats/DataRecord/interface/HcalRecoParamsRcd.h"
+
+#include "Geometry/CaloGeometry/interface/CaloGeometry.h"
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
+#include "Geometry/CaloTopology/interface/HcalTopology.h"
+#include "Geometry/HcalTowerAlgo/interface/HcalGeometry.h"
+
+#include "RecoLocalCalo/HcalRecAlgos/interface/HcalSeverityLevelComputer.h"
+#include "RecoLocalCalo/HcalRecAlgos/interface/HcalSeverityLevelComputerRcd.h"
+#include "RecoLocalCalo/HcalRecAlgos/interface/HcalChannelProperties.h"
+#include "RecoLocalCalo/HcalRecAlgos/interface/HcalChannelPropertiesRecord.h"
+
+class HcalChannelPropertiesEP : public edm::ESProducer {
+public:
+  typedef std::unique_ptr<HcalChannelPropertiesVec> ReturnType;
+
+  inline HcalChannelPropertiesEP(const edm::ParameterSet&) {
+    auto cc = setWhatProduced(this);
+    edm::ESInputTag qTag("", "withTopo");
+    cc.setConsumes(condToken_).setConsumes(topoToken_).setConsumes(paramsToken_);
+    cc.setConsumes(sevToken_).setConsumes(qualToken_, qTag).setConsumes(geomToken_);
+  }
+
+  inline ~HcalChannelPropertiesEP() override {}
+
+  ReturnType produce(const HcalChannelPropertiesRecord& rcd) {
+    // There appears to be no easy way to trace the internal
+    // dependencies of HcalDbService. So, rebuild the product
+    // every time anything changes in the parent records.
+    // This means that we are sometimes going to rebuild the
+    // whole table on the lumi block boundaries instead of
+    // just updating the list of bad channels.
```

Comment:
There is a pattern/mechanism for partial updates, but do I understand correctly that the challenge is not in using that but somewhere else?

Reply:
The problem is that HcalDbService product is an aggregate associated with the record HcalDbRecord which has 25 parent records. Among these parent records, there are records whose IOVs are expected to change on the run boundaries and there records whose IOVs are expected to change on the lumi section boundaries. I don't want to dig into the internals of HcalDbService in order to figure out which pieces of information provided by this service depend on which record, when they get invalidated, etc.